### ISSUE TITLE
Can we expect apache to have a valid asc file on older versions?

### ISSUE THREAD

**Issue by hcpss-banderson** [HCPSS/hub-docker#4]
@hcpss-mbarger,

Tried to build the web image today and it failed:

```
Processing triggers for libc-bin (2.19-18+deb8u1) ...
+ curl -SL https://www.apache.org/dist/httpd/httpd-2.4.17.tar.bz2 -o httpd.tar.bz2
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   229  100   229    0     0    343      0 --:--:-- --:--:-- --:--:--   343
+ curl -SL https://www.apache.org/dist/httpd/httpd-2.4.17.tar.bz2.asc -o httpd.tar.bz2.asc
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   233  100   233    0     0    491      0 --:--:-- --:--:-- --:--:--   492
+ gpg --verify httpd.tar.bz2.asc
gpg: no valid OpenPGP data found.
gpg: the signature could not be verified.
Please remember that the signature file (.sig or .asc)
should be the first file given on the command line.
The command '/bin/sh -c set -x   && apt-get update   && apt-get install -y $systemDeps --no-install-recommends   && apt-get install -y $runtimeDeps --no-install-recommends 	&& apt-get install -y $buildDeps --no-install-recommends   && curl -SL "$HTTPD_BZ2_URL" -o httpd.tar.bz2 	&& curl -SL "$HTTPD_BZ2_URL.asc" -o httpd.tar.bz2.asc 	&& gpg --verify httpd.tar.bz2.asc 	&& mkdir -p src/httpd 	&& tar -xvf httpd.tar.bz2 -C src/httpd --strip-components=1 	&& rm httpd.tar.bz2* 	&& cd src/httpd 	&& ./configure --enable-so --enable-ssl --enable-cgi --enable-rewrite --prefix=$HTTPD_PREFIX --enable-mods-shared=most 	&& make -j"$(nproc)" 	&& make install 	&& cd ../../ 	&& rm -r src/httpd 	&& sed -ri ' 		s!^(\s*CustomLog)\s+\S+!\1 /proc/self/fd/1!g; 		s!^(\s*ErrorLog)\s+\S+!\1 /proc/self/fd/2!g; 		' /usr/local/apache2/conf/httpd.conf' returned a non-zero code: 2
```

I saw there was a new version of Apache out so I bumped the version number from 2.4.17 to 2.4.18 and the image now builds. But my question is, do you think this is normal? Is the asc file missing or invalid in older versions on purpose? Or is this a mistake by Apache? If that is by design then we should refactor to account for it.

**Comment by hcpss-banderson** [comment]
I don't think what we are doing is a way to ensure control over software versions. I have now had a similar issue (#5) with MySQL. We have exact version numbers listed in the docker file (5.7.10-1debian8). We did this under the assumption that it meant we could specify that we wanted version 5.7.10-1debian8 until we chose to update. In reality, it seems that with the current methodology, if we want to build the image, we MUST update to 5.7.11-1debian8 as soon as it is made available in the package manager. This is because once the new version is released, the old 5.7.10-1debian8 is no longer available.

@hcpss-mbarger, what do you think? With the database image it seems like we have lost the convenience of the package manager without gaining the control of compiling the source ourselves.

**Comment by hcpss-mbarger** [comment]
I see now.  In the Apache instance, the older tarball and .asc are moved to an archive folder at http://archive.apache.org/dist/httpd/.  Perhaps we should have a simple wget test that checks for a 200 response on the primary files and, if unsuccessful, runs it using the archive URL.  That way it always pulls the version we're expecting and we only need to rev a version number if we need a security, bug, or performance fix.

**Comment by hcpss-banderson** [comment]
Good idea.

Do you think we should do something similar with MySQL? It would mean moving from the package system to compiling ourselves.

This could also just be an issue with our process. Right now, when we need an image we have to build it. So if we move/add production servers (or testing or staging) and every time a developer brings up a new development environment we build the images. This is particularly painful with the Web image which compiles Apache. When that fails, it derails the process.

We want to lock in the version, but we could just do that by persisting the image and using the same image in all environments. Then it is less important that we be able to build the image at any time because we can always access the built version of the image.

Maybe we can just store exported tar balls of the images:

http://stackoverflow.com/questions/23935141/how-to-copy-docker-images-from-one-host-to-another-without-via-repository

**Comment by hcpss-mbarger** [comment]
I think just saving the tarball is worth a try. I've created a hcpss.web.docker bucket on S3 if you want to try it.

**Comment by hcpss-banderson** [comment]
I just want to post my solution to this problem.

I think if we really want control over the versions of our infrastructure software we should build them into a seperate layer of docker images.

For example let's say we need an apache image that we plan to build and test on 2.4.20 (current as of this writing). We have 3 approaches to getting a working image:

**Comment by hcpss-banderson** [comment]
Accidentally posted half a comment on here if you got a strange email.

**Comment by hcpss-banderson** [comment]
Here's the real comment. I just want to document my solution to this problem.

I think if we really want control over the versions of our infrastructure software we should build them into a separate layer of docker images.

For example let's say we need an apache image that we plan to build and test on 2.4.20 (current as of this writing). We have 2 approaches to getting a working image:

1. Inherit an official image (FROM httpd or with the version FROM httpd:2.4.20) in our own Dockerfile and add our application to a completed image.
2. Inherit from a more primitive base image (debian or ubuntu) and either:
  1. Compile apache ourselves or
  2. Install apache from packages

The downside to 1 is that we don't have much control over how apache is compiled. The official repos usually provide a lot of flexibility for adding extensions etc. But sometimes it isn't enough and to get the configuration you want, you have to install yourself, either from source or from packages.

Which brings us to 2. Now we have complete control over what is installed. But at what cost? For one thing it's slow. If you make a simple code change to your application, you have to wait minutes or hours to build the image. Worse is the errors. It's bad enough that you had to wait for the image to build, but now you have errors and you find yourself spending time troubleshooting compilation or missing package when all you wanted to do was push a small bug fix. I don't like troubleshooting compilation and package problems. Doing so when I just want to commit a code change makes Hulk want to smash.

But wait, there's a 3rd option!

Let's get the best of both worlds by building our own base or intermediate images! Instead of inheriting from an official apache image we build our own. So I install apache 2.4.20 from source or packages and save it as an image called hcpss/apache-for-my-app. Then I inherit that in my application image. Now when I build my application image I know exactly what version of apache I am going to get and how it is configured but I don't have to go through the long and error proned process of installing it.

It might even be beneficial to use more than one intermediate image for a Drupal image your might want:
1. hcpss/apache-for-drupal (`FROM debian:jessie` and apache configured for Drupal)
2. hcpss/drupal (`FROM hcpss/apache-for-drupal` and also drush is installed)
3. hcpss/my-drupal-application (`FROM hcpss/drupal` and the source of my Drupal application)

Then when I am building an image, I can focus on just the thing I am trying to accomplish and not other layers in the application stack.

The only downside is that it can get confusing. When I upgrade hcpss/apache-for-drupal I need to rebuild hcpss/drupal and then hcpss/my-drupal-application.

**Comment by hcpss-mbarger** [comment]
I really appreciate your exploration of these use cases. Love option 3. Best of both worlds indeed.

> When I upgrade hcpss/apache-for-drupal I need to rebuild hcpss/drupal and then hcpss/my-drupal-application.

I'm wondering if switching from an integrated model (mod_php) to a web server (Nginx or Apache) + separate app runtime (PHP-fpm) model would alleviate most of this.  Different processes, different containers and images, using docker to speak across ports. Need to upgrade Nginx for security reasons?  That layer updates while the PHP-fpm layer with the Drupal code remains stable.

**Comment by hcpss-banderson** [comment]
> I'm wondering if switching from an integrated model (mod_php) to a web server (Nginx or Apache) + separate app runtime (PHP-fpm) model would alleviate most of this.

Absolutely. With mod_php, if we want to alter PHP, we also have to build/install Apache. With php-fpm that's not the case.

Then there is the issue of the source code. Since both Nginx and PHP-FPM need access to the source code where should we put it? We can put it in one, the other, neither or both. If you put it in a container with Nginx, then if you upgrade Nginx you will need to recompile the derived image. Same for PHP-FPM. The obvious solution to that is to place the source code in a separate image without Nginx or PHP-FPM but with an exposed volume for both. This minimizes recompilation. The only thing I struggle with is: then in what way are we taking advantage of Docker to ship a working application in a container?

**Comment by hcpss-mbarger** [comment]
It all depends on the application modeling. Thankfully Docker supports the spectrum between using it to ship an application as a single container and using it to power an application composed of multiple services in more than one container.  I'm sure each of our major projects will have natural resting spots along different points on that spectrum.  For instance, my intuition says that the Hub needs to be closer to the former while I'd like to see the main site closer to the latter.

Whatever starting point you decide works for me.

**Comment by hcpss-banderson** [comment]
I'm generally for more segmentation. It just make it confusing to communicate what the structure of the app is. I am about to deploy an application where the instructions for building it are in an ansible deployment playbook. The same instructions are replicated in the dev repo for provisioning a dev environment. It's redundant and hard to understand for a new developer.

I'm going to investigate docker compose. Hopefully it can be a single source of knowledge for the architecture of an application.
